src/codec.rs

```diff
@@ -85,16 +86,27 @@ pub trait Input {
 	/// This is called when decoding reference-based type is finished.
 	fn ascend_ref(&mut self) {}
 
+	/// Hook that is called before allocating memory on the heap.
+	///
+	/// The aim is to get a reasonable approximation of memory usage, especially with variably
+	/// sized types like `Vec`s. Depending on the structure, it is acceptable to be off by a bit.
+	/// For example for structures like `BTreeMap` we don't track the memory  used by the internal
+	/// tree nodes etc. Also we don't take alignment or memory layouts into account.
+	/// But we should always track the memory used by the decoded data inside the type.
+	fn on_before_alloc_mem(&mut self, _size: usize) -> Result<(), Error> {
+		Ok(())
+	}
+
 	/// !INTERNAL USE ONLY!
 	///
-	/// Decodes a `bytes::Bytes`.
+	/// Used when decoding a `bytes::Bytes` from a `BytesCursor` input.
 	#[cfg(feature = "bytes")]
 	#[doc(hidden)]
-	fn scale_internal_decode_bytes(&mut self) -> Result<bytes::Bytes, Error>
+	fn __private_bytes_cursor(&mut self) -> Option<&mut BytesCursor>
```

Comment:
What you are saying there makes no real sense. Decoding `Bytes` from bytes currently doesn't work with any other input. This is by design. So, the changes you have done here are not really required.



> The same applies for combining `BytesCursor` with `DepthLimit` or anything else.

You can simply not do this right now. `BytesCursor` is right now not exposed to the outside and thus, you can also not use it as `input` for `DepthLimit`.

In the end what you would need to do is that you forward the `scale_internal_decode_bytes` call in `DepthLimit` to the inner input for example. Still, this would require that `BytesCursor` is public, which we should not do right now.

Reply:
Hmmm yes, makes sense. Because `BytesCursor` is private. I didn't think of that. Ok, reverted this change then.